src/did.ts

```diff
@@ -0,0 +1,75 @@
+import { DagJson, encodeDagJson } from './dag-json'
+import { RPCProvider } from './rpc'
+
+interface AuthenticateResult {
+  did: string
+}
+
+export interface CreateJWSOptions {
+  protected?: Record<string, any>
+  pubKeyId?: string
+}
+
+interface CreateJWSParams extends CreateJWSOptions {
+  payload: DagJson
+}
+
+interface CreateJWSResult {
+  jws: string // base64-encoded
+}
+
+export class DID {
+  protected _did: string | undefined
+  protected _provider: RPCProvider
+  protected _reqId = 0
+
+  constructor(provider: RPCProvider) {
+    this._provider = provider
+  }
+
+  public get authenticated(): boolean {
+    return this._did != null
+  }
+
+  public get DID(): string {
+    if (this._did == null) {
+      throw new Error('DID is not authenticated')
+    }
+    return this._did
+  }
+
+  protected async _send<P, R>(method: string, params?: P): Promise<R> {
+    const res = await this._provider.send<P, R>({
+      jsonrpc: '2.0',
+      id: `did${this._reqId++}`,
+      method,
+      params,
+    })
+    if (res.error != null) {
+      throw new Error(res.error.message ?? 'RPC error')
+    }
+    return res.result as R
+  }
```

Comment:
@oed I'm wondering if this request envelope logic should live in this class or somewhere else please?
IMO it's a provider/transport layer concern, but if the provider only takes care of implementing the `send: (request: RpcRequest) => Promise<RpcResponse>` interface I guess it has to live here?

Reply:
I think it can live here for now. We have some similar logic in `3id-blockchain-utils` and 3box-js. I'm not sure if there is one best place to put it.